### PR TITLE
feat(step-functions): ternary, join, bug fixes

### LINKED ISSUE #73: Step Functions: gracefully handle empty blocks
The Step Function interpreter does not handle all cases when a `BlockStmt` is empty - it does handle cases like an empty catch (i think), but we need to enumerate the cases thoroughly.

### LINKED ISSUE #78: Step Functions: `undefined` vs `null`
In Step Functions, we currently use `null` in places where we really mean `undefined` in TypeScript. This works for many cases but is definitely a gotcha that could hurt people. How do we model `undefined` in Step Functions?

An object with a missing is "undefined", but how do we return `undefined` as the whole result? We can't - only `null` can be returned.

### LINKED ISSUE #84: Step Functions: chained map, filter, etc.
Optimize chained map/filter calls into single Map tasks with each of the steps inlined to avoid multiple traversals.

```ts
list.map(A).filter(B)
```

Becomes one Map Task
```
Type: Map,
Iterator: {
  A,
  B
}
```

### LINKED ISSUE #210: Step Functions: filters with sub-tasks
```
await Promise.all(input.list.map((item) => task(item)))).filter
-----
Error: .filter with sub-tasks are not yet supported
```

```ts
test("await Promise.all(input.list.map((item) => task(item)))).filter", () => {
  const { stack, task } = initStepFunctionApp();
  const definition = new ExpressStepFunction<
    { list: string[] },
    (number | null)[]
  >(stack, "fn", async (input) => {
    return (await Promise.all(input.list.map((item) => task(item)))).filter(
      (item) => item !== null
    );
  }).definition;

  expect(normalizeDefinition(definition)).toMatchSnapshot();
});
```

### LINKED ISSUE #281: Step Function: Support tunary/conditionalexpr

### PR DESCRIPTION
**Pull request by thantos**
Closes #281 Closes #210 Closes #78 Closes #73
Partial #84

* `join`
* ternary
* bug: out of scope in map states
* removed most of the `skip`s in step function tests.

### CODE REVIEW

File: src/asl.ts [hunk 1]
```diff
@@ -1255,7 +1255,9 @@ export class ASL {
 
     if (isPromiseExpr(expr)) {
       // if we find a promise, ensure it is wrapped in Await or returned then unwrap it
-      if (isAwaitExpr(expr.parent) || isReturnStmt(expr.parent)) {
+      if (
+        expr.hasOnlyAncestors(isConditionExpr, anyOf(isReturnStmt, isAwaitExpr))
```
Comment (sam-goodwin) [inline src/asl.ts:1259]: Does it stop at the first one it finds?

Comment (thantos) [reply src/asl.ts:1259]: The first predicate is a passthrough and the second it stops at

File: src/asl.ts [hunk 2]
```diff
@@ -2099,6 +2164,137 @@ export class ASL {
     }
   }
 
+  private joinToStateOutput(
```
Comment (sam-goodwin) [inline src/asl.ts:2167]: Big function. Could use some documentation

**Review by sam-goodwin — APPROVED** [review]

File: src/asl.ts (hunk 2 above)
Comment (sam-goodwin) [reply src/asl.ts:2167]: I mean at the top. A summary of what syntax it enables. The docs inside are great

**Pull request merged**